File: frontend/src/services/departmentService.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/api/department';

const getAllDepartment = async () => {
    return axios.get(API_URL);
};

const getDepartmentById = async (id) => {
    return axios.get(`${API_URL}/${id}`);
};

const createDepartment = async (department) => {
    return axios.post(API_URL, department);
};

const updateDepartment = async (id, department) => {
    return axios.put(`${API_URL}/${id}`, department);
};

const deleteDepartment = async (id) => {
    return axios.delete(`${API_URL}/${id}`);
};

export {
    getAllDepartment,
    getDepartmentById,
    createDepartment,
    updateDepartment,
    deleteDepartment
}

File: backend/src/main/java/com/example/backend/service/DepartmentService.java
package com.example.backend.service;

import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.backend.dto.DepartmentDTO;
import com.example.backend.entity.Department;
import com.example.backend.repository.ClassSDRepository;
import com.example.backend.repository.DepartmentRepository;

@Service
public class DepartmentService {

    @Autowired
    private DepartmentRepository departmentRepository;

    @Autowired
    private ClassSDRepository classSDRepository;

    public Iterable<Department> getAllDepartment() {
        return departmentRepository.findAll();
    }

    public Optional<Department> getDepartmentById(int id) {
        return departmentRepository.findById(id);
    }

    public Department createDepartment(DepartmentDTO dto) {
        Department department = new Department();
        department.setDepname(dto.getDepname());
        return departmentRepository.save(department);
    }

    public Department updateDepartment(int id, DepartmentDTO dto) {
        Department department = departmentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy"));
        department.setDepname(dto.getDepname());
        return departmentRepository.save(department);
    }

    public boolean deleteDepartment(int id) {
        if (classSDRepository.existsByDepartmentId(id)) {
            throw new RuntimeException("Không thể xóa, khoa này đã có lớp học.");
        }

        if (departmentRepository.existsById(id)) {
            departmentRepository.deleteById(id);
            return true;
        } else {
            return false;
        }
    }
}
